### PR TITLE
contract + connectors (s3, azure, gcs) + duckdb

### PR DESCRIPTION
**Pull request by arnegoeteyn**
Code from Fred

### CODE REVIEW

File: .env
Comment (arnegoeteyn) [inline .env:1]: This should not be committed

File: dv_utils/client.py [hunk 1]
```diff
@@ -16,18 +17,71 @@
 
 class Client:
     """
-    Http client to interact with the Data-village API.
+    Http client to interact with the Datavillage API.
     """
+    DATA_PROVIDER_COLLABORATOR_ROLE_VALUE="DataProvider"
+
+    DATA_PROVIDER_COLLABORATOR_ROLE_VALUE="DataProvider"
+    DATA_CONSUMER_COLLABORATOR_ROLE_VALUE="DataConsumer"
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py]: why is `DATA_PROVIDER_COLLABORATOR_ROLE_VALUE` declared twice? Should one be `CODE_PROVIDER_*`?

File: dv_utils/client.py [hunk 2]
```diff
@@ -16,18 +17,71 @@
...
+        Returns the data descriptor as json 
+        """
+        try:
+            list_participants = self.request(
+                f"/collaborationSpaces/{collaboration_space_id}/collaborators"
+            )
+            if role!=None:
+                filtered_participants = [x for x in list_participants if x["role"]==role]
+            else:
+                filtered_participants = [x for x in list_participants]
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py]: `filtered_participants = [x for x in list_participants if role == None or role == x["role"]]`

File: dv_utils/client.py [hunk 3]
```diff
@@ -16,18 +17,71 @@
...
+        Returns the participants as json 
+        """
+        try:
+            list_participants = self.request(
+                f"/collaborationSpaces/{collaboration_space_id}/collaborators"
+            )
+            if role!=None:
+                filtered_participants = [x for x in list_participants if x["role"]==role]
+            else:
+                filtered_participants=list_participants
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py]: `filtered_participants = [x for x in list_participants if role == None or role == x["role"]]`

File: dv_utils/client.py (hunk 2 above)
Comment (michielrogissart-datavillage) [reply dv_utils/client.py]: even better: `self.get_list_of_participants(collaboration_space_id, role)`

File: dv_utils/client.py [hunk 4]
```diff
@@ -16,18 +17,71 @@
...
+    
+    def get_data_descriptors_for_collaboration_space(self, collaboration_space_id: str, role: str):
+        """
+        Returns the data descriptors as array of json 
+        """
+        list_participants=self.get_list_of_participants(collaboration_space_id,role)
+        if list_participants != None and len(list_participants)>0:
+            for participant in list_participants:
+                client_id=participant["clientId"]
+                data_descriptors=participant.get("dataDescriptors")
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py]: this method isn't returning anything

File: dv_utils/client.py [hunk 5]
```diff
@@ -16,18 +17,71 @@
...
+                f"/collaborationSpaces/{collaboration_space_id}/collaborators"
+            )
+            if role!=None:
+                filtered_participants = [x for x in list_participants if x["role"]==role]
+            else:
+                filtered_participants=list_participants
+            return filtered_participants
+        except Exception as e:
+            logger.error(e)
+            return None
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py:77]: not sure which exceptions you try to handle here, but if it's network-related, maybe we should handle those in the `self.request()` method?

File: dv_utils/client.py [hunk 6]
```diff
@@ -59,7 +113,7 @@ def write_results(self, user_id: str, filename: str, content: str):
             raise Exception("Unsupported result file name: " + filename)
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py]: This was not part of the PR, but I saw this and don't think we want Exceptions to be thrown in a cage but rather catch them and log

File: dv_utils/client.py [hunk 7]
```diff
@@ -16,18 +17,71 @@
...
+            list_participants = self.request(
+                f"/collaborationSpaces/{collaboration_space_id}/collaborators"
+            )
+            if role!=None:
+                filtered_participants = [x for x in list_participants if x["role"]==role]
+            else:
+                filtered_participants=list_participants
+            return filtered_participants
+        except Exception as e:
+            logger.error(e)
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py:42]: should we not use `audit_log(str(e), LogLevel.ERROR)` or something similar?

File: dv_utils/client.py [hunk 8]
```diff
@@ -59,7 +113,7 @@ def write_results(self, user_id: str, filename: str, content: str):
             raise Exception("Unsupported result file name: " + filename)
 
         self.request(
-            f"/clients/{self.settings.client_id}/applications/{self.settings.app_id}/activeUsers/{user_id}/{filename}",
+            f"/clients/{self.settings.collaboration_space_owner_id}/applications/{self.settings.collaboration_space_id}/activeUsers/{user_id}/{filename}",
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py:82]: I don't think this endpoint exists (anymore). I can't find it back in the code.
We should use `/collaborationSpaces/{space_id}/users/{user_id}/appData/{path}` (this checks if the user gave permission)

File: dv_utils/client.py [hunk 9]
```diff
@@ -38,7 +92,7 @@ def get_data(self, user_id: str, rdf_format: str = "turtle"):
         If the format is turtle (default), the returned value is an rdflib graph.
         """
         raw_data = self.request(
-            f"/clients/{self.settings.client_id}/applications/{self.settings.app_id}/activeUsers/{user_id}/data"
+            f"/clients/{self.settings.collaboration_space_owner_id}/applications/{self.settings.collaboration_space_id}/activeUsers/{user_id}/data"
```
Comment (michielrogissart-datavillage) [inline dv_utils/client.py:61]: we are planning on retiring this endpoint. You can keep it for now if everything works, but we are planning to only use `collaborationSpaces/{space_id}/users/{user_id}/appData/{path}`.

The API will prepend the configured `podFolder` from the db to the `path` variable.

File: dv_utils/connectors/azure.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    "location": {
+      "description": "Location of the files on Azure. */* patterns are supported. {model} can be used and will be replaced by the models listed in the data contract.",
+      "required": true
+    },
+    "file_format": {
+      "description": "Format of the files hosted on Azure. csv, json or parquet are supported.",
+      "required": true
+    },
+    "ref_encryption_key": {
+      "description": "Name of the symetric encryption key (or http reference in the Key store - not yet supported) used to by the data provider to encrypte data.",
```
Comment (michielrogissart-datavillage) [inline dv_utils/connectors/azure.json]:
```suggestion
      "description": "Name of the symmetric encryption key (or http reference in the key store - not yet supported) used by the data provider to encrypt data.",
```

File: dv_utils/connectors/azure.json [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    "encryption_key": {
+      "description": "AES 256 symetric encryption key used to encrypt files at rest - works only for parquet files."
+    },
+    "connection_key": {
+      "description": "Connection key id to access the files on Azure.",
+      "required": true
+    },
+    "share_access_token": {
+      "description": "Share access token (alternative solution to connection key) to access the files on Azure",
+      "required": true
```
Comment (michielrogissart-datavillage) [inline dv_utils/connectors/azure.json]: is `share_access_token` required? Can we not have a reasonable default? It looks like a simple boolean. Or should it be `shared_access_token`?

Comment (michielrogissart-datavillage) [reply dv_utils/connectors/azure.json]: I'm guessing it shouldn't be required as it is an 'alternative solution to connection key' and maybe a better name is `shared_access_token`.

File: dv_utils/connectors/azure.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
+import logging
+import copy
+import duckdb
+
+from dv_utils.connectors.connector import Configuration
+from ..log_utils import audit_log, LogLevel
+
+logger = logging.getLogger(__name__)
```
Comment (michielrogissart-datavillage) [inline dv_utils/connectors/azure.py:8]: maybe we should use `audit_log`?

File: dv_utils/connectors/azure.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+    def add_duck_db_connection(self,duckdb,ref_encryption_key: str):
+        description = self.config.description
+        location = self.config.location
+        file_format = self.config.file_format
+        encryption_key=self.config.encryption_key
+        ref_encryption_key = self.config.ref_encryption_key
+
+        connection_key=self.config.connection_key
+        share_access_token=self.config.share_access_token
```
Comment (michielrogissart-datavillage) [inline dv_utils/connectors/azure.py]: of these variables, only `encryption_key`, `ref_encryption_key` and `connection_key` are used. But do we really need to declare them separately if they are only used once?

File: dv_utils/connectors/file.py [hunk 14]
```diff
@@ -4,7 +4,7 @@
 import os
 import copy
 import cloudscraper
-from dv_utils import audit_log, LogLevel
+from ..log_utils import audit_log, LogLevel
```
Comment (michielrogissart-datavillage) [inline dv_utils/connectors/file.py:8]: is this change needed? I think the file connector works, but maybe you had a reason to change this?

File: dv_utils/connectors/s3.json [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+      "required": true
+    },
+    "location": {
+      "description": "Location of the files on S3. */* patterns are supported. {model} can be used and will be replaced by the models listed in the data contract.",
+      "required": true
+    },
+    "file_format": {
+      "description": "Format of the files hosted on S3. csv, json or parquet are supported.",
+      "required": true
+    },
```
Comment (michielrogissart-datavillage) [inline dv_utils/connectors/s3.json:14]: does this mean we use one connector config to download multiple files that all have the same format? Not against it, just wondering if we cannot leverage the extension of the filename to be able to download multiple files in different formats with one connector (if this behaviour makes sense)

File: dv_utils/datasets/contract.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
+"""
+This module define the data contracts for the dv-utils package.
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]:
```suggestion
This module defines the data contracts for the dv-utils package.
```

File: dv_utils/datasets/contract.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        self.data_connector_config_path=data_connector_config_path
+        if data_descriptor != None :
+            self.__init_data_connector(data_descriptor)
+            data_contract_yaml=self.__data_descriptor_to_data_contract(data_descriptor)
+            try:
+                logger.info(f"Create data contract object")
+                data_contract = DataContract(data_contract_str=data_contract_yaml)
+                self.data_contract=data_contract
+            except Exception as inst:
+                logger.error(Exception)
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]: Shouldn't this be `logger.error(inst)` (if we disregard my other comment about using audit_log)?

File: dv_utils/datasets/contract.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+import yaml
+
+from ..settings import Settings
+from ..settings import settings as default_settings
+
+from ..connectors.connector import populate_configuration
+from ..connectors import s3,gcs,azure
+
+from datacontract.data_contract import DataContract
+from datacontract.data_contract import Server
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]: looks like this is not used?

File: dv_utils/datasets/contract.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+            try:
+                logger.info(f"Create data contract object")
+                data_contract = DataContract(data_contract_str=data_contract_yaml)
+                self.data_contract=data_contract
+            except Exception as inst:
+                logger.error(Exception)
+                raise
+        else:  
+            logger.error(f"No data contract available")
+            raise RuntimeError("Unable to create data contract")
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]: same as before: I don't think we want exceptions to be raised in a cage

File: dv_utils/datasets/contract.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+                self.connector = azure.AZConnector(config)
+        except Exception as inst:
+            logger.error(Exception)
+            raise RuntimeError(f"Unable to initialise connector: {inst}")
+
+    def __data_descriptor_to_data_contract(self,data_descriptor: str):
+        try: 
+            if data_descriptor["kind"]=="Dataset":
+                #TODO
+                s=0
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]: can you maybe shortly specify what should happen in this case? 
Also: you can use the keyword `pass` when you don't want to implement it yet

File: dv_utils/datasets/contract.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+'''
+            if data_descriptor["syntheticData"]!="":
+                data_contract_yaml+=f'''
+examples:
+  {data_descriptor["syntheticData"]}
+'''
+            return data_contract_yaml
+        except Exception as inst:
+            logger.error(Exception)
+            raise RuntimeError(f"Unable to transform data descriptor into data contract: {inst}")
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]: same as before

File: dv_utils/datasets/contract.py [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+                data_contract_yaml+=f'''
+models:
+  {data_descriptor["schema"]}
+'''
+            if data_descriptor["syntheticData"]!="":
+                data_contract_yaml+=f'''
+examples:
+  {data_descriptor["syntheticData"]}
+'''
+            return data_contract_yaml
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract.py]: maybe we can use a library like [PyYaml](https://pyyaml.org/wiki/PyYAMLDocumentation) that can serialize a dict to yaml?

File: dv_utils/datasets/contract_quality_check.py [hunk 23]
```diff
@@ -0,0 +1,115 @@
+"""
+This module checks quality of a dataset in a data source for checks defined by the data contract.
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract_quality_check.py]:
```suggestion
This module checks the quality of a dataset in a data source for checks defined by the data contract.
```

File: dv_utils/datasets/contract_quality_check.py [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+  
+    def check_contract_for_data_descriptor(self, collaboration_space_id: str, data_descriptor_id: str,data_connector_config_path:str):
+        logger.info(f"Get data descriptor for collaboration_space_id:{collaboration_space_id} - data_descriptor_id:{data_descriptor_id}")
+        data_descriptor=self.client.get_data_descriptor(collaboration_space_id,data_descriptor_id,self.client.DATA_PROVIDER_COLLABORATOR_ROLE_VALUE)
+        if data_descriptor != None:
+            self.contract.create_contract(data_descriptor,collaboration_space_id,data_descriptor_id,data_connector_config_path)
+            self.__check_contract(self.contract,data_connector_config_path)
+        else:
+            logger.error(f"No data descriptor available for collaboration_space_id:{collaboration_space_id} - data_descriptor_id:{data_descriptor_id}")
+            raise RuntimeError("Unable to check data contract")
```
Comment (michielrogissart-datavillage) [inline dv_utils/datasets/contract_quality_check.py]: no exceptions in cage, or they should be catched whenever the method is used (and maybe should be specified then that it can raise an exception)

File: dv_utils/log_utils.py [hunk 25]
```diff
@@ -7,6 +7,9 @@
 import sys
 from enum import Enum
 
+from .settings import Settings
+from .settings import settings as default_settings
```
Comment (michielrogissart-datavillage) [inline dv_utils/log_utils.py]: As this is the only change to the file, I assume they are not used and thus not needed?

File: dv_utils/settings.py [hunk 26]
```diff
@@ -33,8 +33,8 @@ def load_settings(self, dotenv_file: str = None):
 
         self.base_url: str = config("DV_URL", default="", cast=str)
         self.token: str = config("DV_TOKEN", default="", cast=str)
-        self.app_id: str = config("DV_APP_ID", default="", cast=str)
-        self.client_id: str = config("DV_CLIENT_ID", default="", cast=str)
+        self.collaboration_space_id: str = config("DV_COLLABORATION_SPACE_ID", default="", cast=str)
```
Comment (michielrogissart-datavillage) [inline dv_utils/settings.py]: I don't think this environment variable is injected in the cage. Do you have any particular reason to change this?

File: dv_utils/settings.py [hunk 27]
```diff
@@ -33,8 +33,8 @@ def load_settings(self, dotenv_file: str = None):
...
+        self.collaboration_space_owner_id: str = config("DV_COLLABORATION_SPACE_OWNER_ID", default="", cast=str)
```
Comment (michielrogissart-datavillage) [inline dv_utils/settings.py]: I think this one is also not injected in the cage

**Review by michielrogissart-datavillage — CHANGES_REQUESTED** [review]
Main remarks:
- use audit_log for logging
- don't raise exceptions or add it to the documentation if you do
- rename of env variables might be a problem

File: dv_utils/client.py (hunk 1 above)
Comment (beaule) [reply dv_utils/client.py]: ✅

File: dv_utils/client.py (hunk 2 above)
Comment (beaule) [reply dv_utils/client.py]: ✅ (I've removed the method, not used)

File: dv_utils/client.py (hunk 3 above)
Comment (beaule) [reply dv_utils/client.py]: ✅

File: dv_utils/client.py (hunk 4 above)
Comment (beaule) [reply dv_utils/client.py]: ✅ (I've removed the method, not used)

File: dv_utils/client.py (hunk 6 above)
Comment (beaule) [reply dv_utils/client.py]: ✅

File: dv_utils/client.py (hunk 7 above)
Comment (beaule) [reply dv_utils/client.py:42]: Not really. audit_log is for non repudiation log not technical log (we can discuss this)

File: dv_utils/client.py (hunk 8 above)
Comment (beaule) [reply dv_utils/client.py:82]: This was existing code. I prefer not to touch this :)  But of course please do...

File: dv_utils/client.py (hunk 9 above)
Comment (beaule) [reply dv_utils/client.py:61]: This was existing code. I prefer not to touch this :)  But of course please do...

File: dv_utils/connectors/azure.json (hunk 10 above)
Comment (beaule) [reply dv_utils/connectors/azure.json]: ✅

File: dv_utils/connectors/azure.json (hunk 11 above)
Comment (beaule) [reply dv_utils/connectors/azure.json]: ✅

File: dv_utils/connectors/azure.py (hunk 12 above)
Comment (beaule) [reply dv_utils/connectors/azure.py:8]: Same as before

File: dv_utils/connectors/azure.py (hunk 13 above)
Comment (beaule) [reply dv_utils/connectors/azure.py]: ✅

File: dv_utils/connectors/file.py (hunk 14 above)
Comment (beaule) [reply dv_utils/connectors/file.py:8]: It was to align on others

File: dv_utils/connectors/s3.json (hunk 15 above)
Comment (beaule) [reply dv_utils/connectors/s3.json:14]: usually it's all in the same format. I can explain

File: dv_utils/datasets/contract.py (hunk 16 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: ✅

File: dv_utils/datasets/contract.py (hunk 17 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: ✅ changed (for the audit log, same as before)

File: dv_utils/datasets/contract.py (hunk 18 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: ✅

File: dv_utils/datasets/contract.py (hunk 19 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: ✅

File: dv_utils/datasets/contract.py (hunk 20 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: ✅

File: dv_utils/datasets/contract.py (hunk 21 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: ✅

File: dv_utils/datasets/contract.py (hunk 22 above)
Comment (beaule) [reply dv_utils/datasets/contract.py]: Need to discuss :)

File: dv_utils/datasets/contract_quality_check.py (hunk 23 above)
Comment (beaule) [reply dv_utils/datasets/contract_quality_check.py]: ✅

File: dv_utils/datasets/contract_quality_check.py (hunk 24 above)
Comment (beaule) [reply dv_utils/datasets/contract_quality_check.py]: ✅

File: dv_utils/log_utils.py (hunk 25 above)
Comment (beaule) [reply dv_utils/log_utils.py]: ✅

File: dv_utils/settings.py (hunk 26 above)
Comment (beaule) [reply dv_utils/settings.py]: Need to discuss

File: dv_utils/settings.py (hunk 27 above)
Comment (beaule) [reply dv_utils/settings.py]: need to discuss

File: dv_utils/log_utils.py (hunk 25 above)
Comment (michielrogissart-datavillage) [reply dv_utils/log_utils.py]: looks like you forgot to delete this one but it's ok, I will do it
